analytics_test.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// TestAddWeeklyUsageWithoutAnyData tests adding a new daily usage into the weekly usage if
+// it is not existed
+func TestAddWeeklyUsageWithoutAnyData(t *testing.T) {
+	now := time.Now()
+	u := WeeklyUsage{
+		Data: map[time.Weekday]DailyUsage{},
+	}
+	u.CountRequests(5)
+
+	assert.Equal(t, 5, u.Data[now.Weekday()][now.Hour()])
+}
+
+// TestAddWeeklyUsageWithOldData tests if the weekly usage contains the data of
+// the same weekday, it will be added up
+func TestAddWeeklyUsageWithOldData(t *testing.T) {
+	now := time.Now()
+	u := WeeklyUsage{
+		Data: map[time.Weekday]DailyUsage{},
+	}
+	u.Data[now.Weekday()] = DailyUsage{
+		now.Hour(): 3,
+	}
+	u.CountRequests(5)
+
+	assert.Equal(t, 8, u.Data[now.Weekday()][now.Hour()])
+}
+
+// TestAddWeeklyUsageAcrossNextDay tests that the daily usage data
+// will be removed if the usage data comes from a new day
+func TestAddWeeklyUsageAcrossNextDay(t *testing.T) {
+	now := time.Now()
+	u := WeeklyUsage{
+		Today: now.Add(-24 * time.Hour),
+		Data: map[time.Weekday]DailyUsage{
+			// yesterday's usage
+			now.Weekday() - 1: {
+				now.Hour(): 2,
+			},
+			// today's usage
```

Comment:
this one should be the same weekday last week?

Reply:
I do not understand what you mean. This counts usages day-by-day.

Reply:
Besides, yesterday's record is not really required this test case. I will remove it.

Reply:
Let's say today is Wednesday. `map[now.Weekday()]` is going to record the usage of today. The current value `{now.Hour(): 3}` should be the usage of last Wednesday and that's why the value will be replaced with today's data. I was confused because the data here doesn't reflect today's usage yet.

Reply:
Last Wednesday's data should be remove when the very first request of today arrives so it should reflect today's usage.

Reply:
If you are worrying that no request causes the old usage is not cleaned up, there will be another function to do periodical check to make sure the old data will be always clean. It is not relevant to this test case. The test case is to make sure the very first data of a day will also create a refresh record instead of using the old one.